### PR TITLE
Basic support for braille display key modifiers

### LINKED ISSUE #6213: Use braille display keys as modifier keys
The feature I propose is best described with the following example.

There are several braille displays with function keys assigned to specific keyboard modifier keys, one of which is the Hims Braille Edge. This display has 8 function keys, from left to right meant to be assigned to Esc, Tab, Ctrl, Alt, Shift, Insert, Windows, and the application key. This display also has two circular groups of four buttons, meant to be arrow keys.

The mentioned braille display function keys are especially handy if you combine them with other function keys or arrow keys. So when you press advance4 (ctrl) + advance1 (tab), you should be able to simulate the ctrl+tab gesture on the keyboard.

IN the current situation, every single combination has to be manually defined in NVDA. Thus, if you assign br(hims):advance3 to kb:control and br(hims):leftSideLeftArrow to kb:leftArrow, the combination advance3+leftSideLeftArrow isn't automatically assigned to kb:control+leftArrow, unless you specifically do this, either manually or in the driver itself. Especially the in-driver assignments consume a lot of code and are error-prone, since both the keyboard key and braille key assignments reside in strings and there is currently no meganism to test them other than with the braille display in front of you.

This leads to the idea of adding a way to use braille display keys assigned to keyboard modifier keys as modifier keys themselves. In the case of Hims, if advance3 is assigned to ctrl and advance4 to alt, br(hims):advance3+advance4+leftSideLeftArrow should sum up to kb:ctrl+alt+leftArrow without the need of manually assigning this combination. Using these modifier keys in other composite key strokes (containing arrow key emulators or even braille input combinations) should be possible without manual assignments. This would open great possibilities, pressing key strokes like ctrl+c, ctrl+v, ctrl+space without keyboard intervention and manual assignments. 

I understand that this proposal can conflict with current key assignments in certain display drivers, which is a point to discuss. I see the following options:
1. Always give assigned combinations precedense. When advance3=ctrl and leftSideLeftArrow=leftArrow, it should be possible to assign advance3+leftSideLeftArrow to something entirely different, like moving the review cursor left
2. Ignore those assigned combinations. If advance3+leftSideLeftArrow is assigned for moving the review cursor left, that assignment should be ignored and the combination should consistently simulate ctrl+leftArrow

In other words, point 1 stands for freedom, point 2 stands for concistency. Since I see reasons why both options could have preference, I would vote for making this an option in the braille settings dialog

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Fixes #6213 

### Description
This adds basic support for modifiers in braille display gestures that emulate keyboard gestures. For example, if you have baum key b10 assigned to kb:control and you combine b10 with the joystick found on baum displays, you will be able to perform keystrokes like control+upArrow, control+downArrow, etc. without explicitly defining them.

### Testing performed:
Tested the example as noted above.

### Known issues with pull request:
* Combining modifiers with braille input (#7306) is not yet supported. That's why I call this basic support for modifiers.
* This does not yet have unit tests. Since we actually have a braille module in unit tests, it should be possible to provide them. However, I'd like to have the current approach reviewed first, if possible.

### CODE REVIEW

File: source/braille.py [hunk 1]
```diff
@@ -1925,6 +1926,25 @@ def getPossiblePorts(cls):
 	#: @type: L{inputCore.GlobalGestureMap}
 	gestureMap = None
 
+	@classmethod
+	def _getModifierGestures(cls):
+		"""Retrieves modifier gestures from this display driver's L{gestureMap}
+		that are bound to modifier only keyboard emulate scripts.
+		@return: the ids of the display keys and the associated generalised modifier names
+		@rtype: generator of (set, set)
```
Comment (dkager) [inline source/braille.py]: Why sets instead of tuples/lists? Sets have some fancy operator overloads, but that pattern is not used much in NVDA.

Comment (leonardder) [reply source/braille.py]: I use sets because of these operator overloads.

File: source/braille.py [hunk 2]
```diff
@@ -1974,6 +1994,49 @@ def _get_scriptableObject(self):
...
+			return self.script
+		# No script for this gesture has been found, so process this gesture for possible modifiers. 
+		# For example, if L{self.id} is 'key1+key2',
+		# key1 is bound to 'kb:control' and key2 to 'kb:tab',
+		# this gesture should execute 'kb:control+tab'.
+		# Combining modifiers with braille input (#7306) is not yet supported.
+		gestureKeys = set(self.keyNames)
+		gestureModifiers = set()
+		for keys, modifiers in handler.display._getModifierGestures():
+			if keys<gestureKeys:
```
Comment (leonardder) [inline source/braille.py]: @dkager: see this line and lines below, which contain some set operator stuff.

Comment (dkager) [reply source/braille.py]: I assume `<` means "subset of", but I'm not sure. That's why I am wondering if the benefit of overloaded operators is really a benefit.

Comment (leonardder) [reply source/braille.py]: Exactly.

**Review by michaelDCurran — CHANGES_REQUESTED** [review]

File: source/scriptHandler.py [hunk 3]
```diff
@@ -36,6 +36,7 @@ def _makeKbEmulateScript(scriptName):
 		# __name__ must be str; i.e. can't be unicode.
 		scriptName = scriptName.encode("mbcs")
 	func.__name__ = "script_%s" % scriptName
+	func.emuGesture = emuGesture
```
Comment (michaelDCurran) [inline source/scriptHandler.py]: I don't think this is needed anymore?

**Review by michaelDCurran — CHANGES_REQUESTED** [review]

**Review by michaelDCurran — APPROVED** [review]

**Comment by michaelDCurran** [comment]
Is this ready to be incubated?

**Comment by leonardder** [comment]
Sure.

**Comment by michaelDCurran** [comment]
@leonardder: could you please come up with an entry for the changes for developers section of what's new? Then I can merge this to master. Something referring to the fact it is no longer necessary to bind individual key combinations in braille display drivers...

**Comment by leonardder** [comment]
I usually do not provide changes entries as my English grammar is not what it should be, but I will be more complete for new pull requests. How about the following:

* Braille display gestures that emulate system keyboard key modifiers (such as control and alt) can now be combined with other emulated system keyboard keys without explicit definition.
    + For example: if you have a key on your display bound to the alt key and another display key to downArrow, combining these keys will result in the emulation of alt+downArrow.

**Comment by michaelDCurran** [comment]
Looks good, thank you :)

**Pull request merged**